admin/main.go

```diff
@@ -138,3 +150,77 @@ func main() {
 		os.Exit(1)
 	}
 }
+
+// findSocketOrPort detects if pmm-agent is listening on socket or port.
+// The arguments are used for testing purposes and can be left empty.
+func findSocketOrPort(socketPath string, localPort uint32) (string, uint32) {
+	logrus.Debug("Detecting socket or port of local pmm-agent")
+
+	if socketPath == "" {
+		socketPath = flags.SocketPath
+	}
+
+	if localPort == 0 {
+		localPort = agentlocal.DefaultPMMAgentListenPort
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	// We make the channels buffered with size 1 to avoid goroutine leaks
+	socketChan := make(chan string, 1)
+	portChan := make(chan uint32, 1)
+
+	// Check socket connection
+	go func() {
+		defer close(socketChan)
+
+		var socket string
+
+		dialer := net.Dialer{}
+		conn, err := dialer.DialContext(ctx, "unix", socketPath)
+		if err == nil {
+			err := conn.Close()
+			if err != nil {
+				logrus.Debugf("Socket close error: %#v", err)
+			}
+
+			socket = socketPath
+		}
+
+		socketChan <- socket
+	}()
+
+	// Check TCP connection
+	go func() {
+		defer close(portChan)
+
+		var port uint32
+
+		dialer := net.Dialer{}
+		conn, err := dialer.DialContext(ctx, "tcp", net.JoinHostPort(agentlocal.Localhost, strconv.Itoa(int(localPort))))
+		if err == nil {
+			err := conn.Close()
+			if err != nil {
+				logrus.Debugf("TCP close error: %#v", err)
+			}
+
+			port = localPort
+		}
+
+		portChan <- port
+	}()
+
+	if sock := <-socketChan; sock != "" {
+		logrus.Debugf("Found socket %s", sock)
+		return sock, 0
+	}
+
+	if port := <-portChan; port != 0 {
+		logrus.Debugf("Found port %d", port)
+		return "", port
+	}
+
```

Comment:
It looks like this code will wait a value from the `socketChan` until it appears and never will go down and check the `portChan`. Is it intended? The [simplified test](https://go.dev/play/p/n55giKb85g1) to play around

Reply:
In your example that's true. In this case though, `socketChan` can receive an empty string, in which case `portChan` message is processed.

Reply:
Oh, I see, you are right)